File: customer-support-ai/server/src/services/chatService.ts
import { v4 as uuidv4 } from 'uuid';
import { ollamaService } from './ollamaService.js';
import { vectorStore } from './vectorStore.js';
import { documentService } from './documentService.js';
import { logger } from '../utils/logger.js';
import type { ChatSession, ChatMessage, ChatRequest, ChatResponse, Source } from '../types/index.js';

// In-memory chat sessions (replace with Redis/DB in production)
const sessions: Map<string, ChatSession> = new Map();

const SYSTEM_PROMPT = `You are a helpful Customer Support AI Assistant. Your role is to answer user questions based ONLY on the provided company documentation and knowledge base.

Guidelines:
1. Answer questions using ONLY the information from the provided context
2. If the answer is not in the context, respond with: "I'm sorry, I don't have information about that in our knowledge base. Would you like me to escalate this to a human agent?"
3. Be concise but thorough in your responses
4. Maintain a professional, friendly tone
5. If asked about something unrelated to customer support, politely redirect to support topics
6. Cite the source document when possible
7. Do not make up information or hallucinate answers

Context from knowledge base:`;

export class ChatService {
  async processMessage(request: ChatRequest): Promise<ChatResponse> {
    const { message, sessionId } = request;
    const currentSessionId = sessionId || uuidv4();

    // Get or create session
    let session = sessions.get(currentSessionId);
    if (!session) {
      session = {
        id: currentSessionId,
        messages: [],
        createdAt: new Date(),
        updatedAt: new Date()
      };
      sessions.set(currentSessionId, session);
    }

    // Add user message
    const userMessage: ChatMessage = {
      id: uuidv4(),
      role: 'user',
      content: message,
      timestamp: new Date()
    };
    session.messages.push(userMessage);

    // Search knowledge base
    logger.info(`🔍 Searching knowledge base for: "${message.substring(0, 50)}..."`);
    const searchResults = await vectorStore.search(message, 5);

    // Enrich sources with document names
    const sources: Source[] = [];
    for (const result of searchResults) {
      const doc = await documentService.getDocument(result.documentId);
      if (doc) {
        sources.push({
          ...result,
          documentName: doc.originalName
        });
      }
    }

    // Build context from sources
    const context = sources.length > 0
      ? sources.map((s, i) => `[Source ${i + 1} from ${s.documentName}]: ${s.chunkContent}`).join('\n\n')
      : 'No relevant documents found in the knowledge base.';

    // Build messages for LLM
    const messagesForLLM = [
      {
        role: 'system',
        content: `${SYSTEM_PROMPT}\n\n${context}`
      },
      ...session.messages.slice(-10).map(m => ({
        role: m.role,
        content: m.content
      }))
    ];

    // Generate response
    logger.info('🤖 Generating AI response...');
    let responseContent: string;

    try {
      responseContent = await ollamaService.chat(messagesForLLM);
    } catch (error) {
      logger.error('LLM generation failed:', error);
      responseContent = "I'm sorry, I'm having trouble processing your request right now. Please try again in a moment.";
    }

    // Add assistant message
    const assistantMessage: ChatMessage = {
      id: uuidv4(),
      role: 'assistant',
      content: responseContent,
      timestamp: new Date(),
      sources: sources.length > 0 ? sources : undefined
    };
    session.messages.push(assistantMessage);
    session.updatedAt = new Date();

    logger.info('✅ Response generated');

    return {
      message: assistantMessage,
      sessionId: currentSessionId,
      sources: sources.length > 0 ? sources : undefined
    };
  }

  async *streamMessage(request: ChatRequest): AsyncGenerator<{ type: string; data?: any }> {
    const { message, sessionId } = request;
    const currentSessionId = sessionId || uuidv4();

    let session = sessions.get(currentSessionId);
    if (!session) {
      session = {
        id: currentSessionId,
        messages: [],
        createdAt: new Date(),
        updatedAt: new Date()
      };
      sessions.set(currentSessionId, session);
    }

    // Add user message
    const userMessage: ChatMessage = {
      id: uuidv4(),
      role: 'user',
      content: message,
      timestamp: new Date()
    };
    session.messages.push(userMessage);

    // Search knowledge base
    yield { type: 'status', data: { status: 'searching' } };
    const searchResults = await vectorStore.search(message, 5);

    const sources: Source[] = [];
    for (const result of searchResults) {
      const doc = await documentService.getDocument(result.documentId);
      if (doc) {
        sources.push({
          ...result,
          documentName: doc.originalName
        });
      }
    }

    if (sources.length > 0) {
      yield { type: 'sources', data: { sources } };
    }

    // Build context
    const context = sources.length > 0
      ? sources.map((s, i) => `[Source ${i + 1} from ${s.documentName}]: ${s.chunkContent}`).join('\n\n')
      : 'No relevant documents found in the knowledge base.';

    const messagesForLLM = [
      {
        role: 'system',
        content: `${SYSTEM_PROMPT}\n\n${context}`
      },
      ...session.messages.slice(-10).map(m => ({
        role: m.role,
        content: m.content
      }))
    ];

    // Stream response
    yield { type: 'status', data: { status: 'generating' } };

    let fullResponse = '';
    const messageId = uuidv4();

    try {
      for await (const chunk of ollamaService.streamChat(messagesForLLM)) {
        fullResponse += chunk;
        yield { type: 'chunk', data: { chunk, messageId } };
      }
    } catch (error) {
      logger.error('Streaming failed:', error);
      const errorMsg = "I'm sorry, I'm having trouble right now. Please try again.";
      yield { type: 'chunk', data: { chunk: errorMsg, messageId } };
      fullResponse = errorMsg;
    }

    // Save assistant message
    const assistantMessage: ChatMessage = {
      id: messageId,
      role: 'assistant',
      content: fullResponse,
      timestamp: new Date(),
      sources: sources.length > 0 ? sources : undefined
    };
    session.messages.push(assistantMessage);
    session.updatedAt = new Date();

    yield { type: 'done', data: { messageId, sessionId: currentSessionId } };
  }

  async getSession(sessionId: string): Promise<ChatSession | undefined> {
    return sessions.get(sessionId);
  }

  async getAllSessions(): Promise<ChatSession[]> {
    return Array.from(sessions.values()).sort(
      (a, b) => b.updatedAt.getTime() - a.updatedAt.getTime()
    );
  }

  async deleteSession(sessionId: string): Promise<boolean> {
    return sessions.delete(sessionId);
  }
}

export const chatService = new ChatService();

File: python_backend/app/chat_service.py
import uuid
from datetime import datetime
from app.config import MAX_CONTEXT_MESSAGES
from app.models import ChatSession, ChatMessage, ChatRequest, ChatResponse, Source
from app.ollama_service import ollama_service
from app.vector_store import vector_store
from app.document_service import document_service
from app.logger import logger

NO_INFORMATION_RESPONSE = "I'm sorry, I don't have information about that in our knowledge base. Would you like me to escalate this to a human agent?"

SYSTEM_PROMPT = f"""You are a helpful Customer Support AI Assistant. Your role is to answer user questions based ONLY on the provided company documentation and knowledge base.

Guidelines:
1. Answer questions using ONLY the information from the provided context
2. If the answer is not in the context, respond with: "{NO_INFORMATION_RESPONSE}"
3. Be concise but thorough in your responses
4. Maintain a professional, friendly tone
5. If asked about something unrelated to customer support, politely redirect to support topics
6. Cite the source document when possible
7. Do not make up information or hallucinate answers

Context from knowledge base:"""


class ChatService:
    def __init__(self):
        self.sessions: dict[str, ChatSession] = {}

    async def process_message(self, request: ChatRequest) -> ChatResponse:
        message = request.message
        session_id = request.session_id or str(uuid.uuid4())

        session = self.sessions.get(session_id)
        if not session:
            session = ChatSession(id=session_id)
            self.sessions[session_id] = session

        user_message = ChatMessage(role="user", content=message)
        session.messages.append(user_message)

        logger.info(f"Searching knowledge base for: \"{message[:50]}...\"")
        search_results = await vector_store.search(message, 5)

        sources: list[Source] = []
        for result in search_results:
            doc = await document_service.get_document(result.document_id)
            if doc:
                result.document_name = doc.original_name
                sources.append(result)

        if not sources:
            logger.info("No relevant sources found - returning escalation message")
            assistant_message = ChatMessage(
                role="assistant",
                content=NO_INFORMATION_RESPONSE,
                sources=None,
            )
            session.messages.append(assistant_message)
            session.updated_at = datetime.utcnow()
            return ChatResponse(
                message=assistant_message,
                session_id=session_id,
                sources=None,
            )

        context = "\n\n".join(
            f"[Source {i + 1} from {s.document_name}]: {s.chunk_content}"
            for i, s in enumerate(sources)
        )

        messages_for_llm = [
            {"role": "system", "content": f"{SYSTEM_PROMPT}\n\n{context}"},
        ]

        for msg in session.messages[-MAX_CONTEXT_MESSAGES:]:
            messages_for_llm.append({"role": msg.role, "content": msg.content})

        logger.info("Generating AI response...")
        try:
            response_content = await ollama_service.chat(messages_for_llm)
        except Exception as e:
            logger.error(f"LLM generation failed: {e}")
            response_content = "I'm sorry, I'm having trouble processing your request right now. Please try again in a moment."

        assistant_message = ChatMessage(
            role="assistant",
            content=response_content,
            sources=sources if sources else None,
        )
        session.messages.append(assistant_message)
        session.updated_at = datetime.utcnow()

        logger.info("Response generated")

        return ChatResponse(
            message=assistant_message,
            session_id=session_id,
            sources=sources if sources else None,
        )

    async def stream_message(self, request: ChatRequest):
        message = request.message
        session_id = request.session_id or str(uuid.uuid4())

        session = self.sessions.get(session_id)
        if not session:
            session = ChatSession(id=session_id)
            self.sessions[session_id] = session

        user_message = ChatMessage(role="user", content=message)
        session.messages.append(user_message)

        yield {"type": "status", "data": {"status": "searching"}}

        search_results = await vector_store.search(message, 5)

        sources: list[Source] = []
        for result in search_results:
            doc = await document_service.get_document(result.document_id)
            if doc:
                result.document_name = doc.original_name
                sources.append(result)

        if sources:
            yield {"type": "sources", "data": {"sources": [s.model_dump() for s in sources]}}

        if not sources:
            logger.info("No relevant sources found - returning escalation message")
            yield {"type": "done", "data": {"message_id": str(uuid.uuid4()), "session_id": session_id, "content": NO_INFORMATION_RESPONSE}}
            assistant_message = ChatMessage(
                id=str(uuid.uuid4()),
                role="assistant",
                content=NO_INFORMATION_RESPONSE,
                sources=None,
            )
            session.messages.append(assistant_message)
            session.updated_at = datetime.utcnow()
            return

        context = "\n\n".join(
            f"[Source {i + 1} from {s.document_name}]: {s.chunk_content}"
            for i, s in enumerate(sources)
        )

        messages_for_llm = [
            {"role": "system", "content": f"{SYSTEM_PROMPT}\n\n{context}"},
        ]
        for msg in session.messages[-MAX_CONTEXT_MESSAGES:]:
            messages_for_llm.append({"role": msg.role, "content": msg.content})

        yield {"type": "status", "data": {"status": "generating"}}

        full_response = ""
        message_id = str(uuid.uuid4())

        try:
            async for chunk in ollama_service.stream_chat(messages_for_llm):
                full_response += chunk
                yield {"type": "chunk", "data": {"chunk": chunk, "message_id": message_id}}
        except Exception as e:
            logger.error(f"Streaming failed: {e}")
            error_msg = "I'm sorry, I'm having trouble right now. Please try again."
            yield {"type": "chunk", "data": {"chunk": error_msg, "message_id": message_id}}
            full_response = error_msg

        assistant_message = ChatMessage(
            id=message_id,
            role="assistant",
            content=full_response,
            sources=sources if sources else None,
        )
        session.messages.append(assistant_message)
        session.updated_at = datetime.utcnow()

        yield {"type": "done", "data": {"message_id": message_id, "session_id": session_id}}

    async def get_session(self, session_id: str) -> ChatSession | None:
        return self.sessions.get(session_id)

    async def get_all_sessions(self) -> list[ChatSession]:
        sessions = list(self.sessions.values())
        sessions.sort(key=lambda s: s.updated_at, reverse=True)
        return sessions

    async def delete_session(self, session_id: str) -> bool:
        if session_id in self.sessions:
            del self.sessions[session_id]
            return True
        return False


chat_service = ChatService()
